File: src/inPython/inheritance/Inheritance3.py
# package python.inheritance

# Eliminating redundant code using an abstract base class
# and letting base class implement interface (i.e. all sub classes
# must implement)
from abc import ABC, abstractmethod


def inheritance_program():
    # pet = Pet("any", 0)     # No longer allowed, Pet is abstract

    dog = Dog("Lassie", 4)
    cat = Cat("Misse", 5, False)

    print(isinstance(dog, Pet))
    print(isinstance(dog, CanSpeak))
    print(isinstance(dog, Dog))
    print(type(dog))

    # Use overridden methods
    print(dog.say())
    print(cat.say())


class CanSpeak(ABC):
    @abstractmethod
    def say(self):
        raise NotImplementedError


# Pet doesn't need to implement "say" since it's marked abstract (inherits from ABC)
class Pet(CanSpeak, ABC):
    def __init__(self, name: str, age: int):
        self.name = name  # Everything common for Cat and Dog here
        self.age = age

    def get_name(self):
        return self.name

    def get_age(self):
        return self.age


class Dog(Pet):
    # def __init__(self, name: str, age: int):
    #     super().__init__(name, age)

    # Must implement say() since Dog is a non-abstract class
    def say(self):
        return f"{self.name} says Arf"

    # getName() and getAge() inherited but invisible


class Cat(Pet):
    def __init__(self, name: str, age: int, is_evil: bool):
        super().__init__(name, age)
        self.is_evil = is_evil

    # Must implement say() since Cat is a non-abstract class
    def say(self):
        return f"{self.name} says Meow"

    # get_name() and get_age() inherited but invisible

    def is_evil(self):
        return self.is_evil


if __name__ == "__main__":
    inheritance_program()

File: src/inJava/inheritance/Inheritance3.java
package inJava.inheritance;

import static java.lang.System.out;

// Eliminating redundant code using an *abstract* common superclass,
// letting superclass define the common interface that all subclasses
// must then implement.
// Note: I use interface in the general sense here, as in, the methods
// and attributes associated with the type; and not in the sense of the
// Java interface mechanism/keyword.
class Inheritance3 {
    static void inheritanceProgram() {
        // Pet _pet = new Pet("any", 0); // No longer allowed, Pet is abstract
        Dog dog = new Dog("Lassie",4);
        Cat cat = new Cat("Misse",5,false);

        out.println(dog instanceof Pet);
        out.println(dog instanceof CanSpeak);
        out.println(dog instanceof Dog);
        out.println(dog.getClass());

        // Use duplicated methods
        out.println(dog.say());
        out.println(cat.say());
    }

    static interface CanSpeak {
        String say();
    }

    // Pet doesn't need to implement "say" since it's
    // marked abstract.
    static abstract class Pet implements CanSpeak {
        private String name;
        private int age;

        Pet(String name, int age){
            this.name = name;
            this.age = age;
        }

        String getName() {
            return name;
        }

        int getAge() {
            return age;
        }

        // This isn't actually needed, the signature is inherited
        // from CanSpeak; I'm just putting it here to make it explicit
        // to you that there's an abstract method in the class.
        @Override
        public abstract String say();
    }

    static class Dog extends Pet {
        Dog(String name, int age) {
            super(name, age);
        }

        @Override
        public String say() {
            return this.getName() + " says Arf";
        }

        // getName() and getAge() inherited but invisible
    }

    static class Cat extends Pet {
        private boolean isEvil;

        Cat(String name, int age, boolean isEvil) {
            super(name, age);
            this.isEvil = isEvil;
        }

        boolean isEvil() {
            return this.isEvil;
        }

        // getName() and getAge() inherited but invisible

        @Override
        public String say() {
            return this.getName() + " says Meow";
        }
    }
    public static void main(String[] args) {
        inheritanceProgram();
    }
}
